.drone.yml

```diff
@@ -3156,6 +3156,90 @@ steps:
     commands:
     - aws s3 sync /rpmrepo/teleport/ s3://$AWS_S3_BUCKET/teleport/
 
+  - name: Download DEB repo contents
+    image: amazon/aws-cli
+    environment:
+      AWS_S3_BUCKET:
+        from_secret: DEBREPO_AWS_S3_BUCKET
+      AWS_ACCESS_KEY_ID:
+        from_secret: DEBREPO_AWS_ACCESS_KEY_ID
+      AWS_SECRET_ACCESS_KEY:
+        from_secret: DEBREPO_AWS_SECRET_ACCESS_KEY
+    volumes:
+      - name: debrepo
+        path: /debrepo
+    commands:
+    # we explicitly want to delete anything present locally which has been deleted
+    # from the upstream S3 bucket
+    - aws s3 sync s3://$AWS_S3_BUCKET/teleport /debrepo/teleport --delete
+    - mkdir -p /debrepo/teleport
+
+  - name: Build DEB repo
+    image: ubuntu:20.04
+    environment:
+      DEBIAN_FRONTEND: noninteractive
+      GNUPGHOME: /tmpfs/gnupg
+      GPG_RPM_SIGNING_ARCHIVE:
+        from_secret: GPG_RPM_SIGNING_ARCHIVE
+    volumes:
+      - name: dockersock
+        path: /var/run
+      - name: debrepo
+        path: /debrepo
+      # for in-memory tmpfs for key material
+      - name: tmpfs
+        path: /tmpfs
+    commands:
+      - |
+        # install needed tools
+        apt-get -y update && apt-get -y install curl gzip gnupg2 reprepro tar
+      - |
+        # write config files
+        mkdir -p /go/reprepro/teleport/conf /go/reprepro/teleport/public
+        cat << EOF > /go/reprepro/teleport/conf/distributions
+        Origin: teleport
+        Label: teleport
+        Codename: stable
+        Architectures: i386 amd64
+        Components: main
+        Description: apt repository for teleport
+        SignWith: 6282C411
```

Comment:
can we somehow extract the key ID from `$GNUPGHOME` after extracting it?

Reply:
Theoretically we could do this with some shell magic, but implementing it this way is actually a bit of a safeguard.

The Drone config can't be updated without being signed, and it can't be signed by anyone who isn't authorized to do so. This means that if the secret containing the key were somehow to be changed, the key's fingerprint would change and then signing will fail, thus notifying us that something is wrong. This also provides a little bit of public transparency that we're actually signing things with the same key that that we're publishing the fingerprint for.

Reply:
Gotcha, so it basically forces a second pair of eyes on key changes.
At some point in the future, we will want to start periodically rotating all of our signing keys. It'll be great to document the process for that somewhere.